### ISSUE TITLE
Clearly define our browser support policy, including which browser + screen reader combinations we support

### ISSUE THREAD

**Issue by claviska** [shoelace-style/webawesome#146]
We need to formalize this because screen readers are finicky. Our decision should be based on the most common and reliable combinations. Bugs in other combos can be addressed on an issue-per-issue basis.

As a small team, we unfortunately can't afford spending too much time on edge cases that only impact a small number of users.

**Comment by lindsaym-fa** [comment]
WebAIM's 2023-24 screen reader survey shows JAWS with Chrome as most common, followed by NVDA with Chrome: https://webaim.org/projects/screenreadersurvey10/#browsercombos

If we want to avoid any commitments to JAWS ([NVDA is more common overall than JAWS at this point](https://webaim.org/projects/screenreadersurvey10/#used)), I'd vote for verifying the following:

- NVDA with Chrome
- NVDA with Firefox
- VoiceOver with Safari

That said, I wouldn't mind _someone_ on the team having JAWS access. It has significant usage and was a favorite among a lot of the visually impaired users I used to work with.

**Comment by KonnorRogers** [comment]
I think @lindsaym-fa is correct.

These combos aren't set in stone, but I do think it's a solid start and voice my support for:

 - NVDA with Chrome
 - NVDA with Firefox
 - VoiceOver with Safari

That's not to say we won't support other combos, but these are going to be our main permutations that we test. If someone raises an issue or submits a PR to support a different combo and it doesn't affect these main combinations, I think it'd be fine to add them

Obviously we'd love to add JAWS, but I know licenses can be expensive and a pain to get.

We also haven't touched on mobile devices either, but perhaps thats a discussion for another time.

**Comment by claviska** [comment]
This seems like a consensus. I'll get the docs updated soon!

**Comment by claviska** [comment]
Fixed in #958

**Comment by myfonj** [comment]
Along with supported browsers and accessibility tools, maybe consider also mentioning what settings and environments is this ~fra~_library_ intended for, namely its relation to

### Progressive enhancement / Graceful degradation

or the lack (??) thereof. It would potentially save developers some time clicking around and figuring out on their own.

(Sorry for being late to the party; I guess the idea is clear to everyone involved here, but for sake of completeness, let me do advocacy.)

I think it would be good for developers to learn about the philosophy of this project with regard to the place of rendering and resilience in non-standard system environment conditions and application settings both in reportedly supported browsers (such as flaky connections, problems with JS, blocked JS, JS turned off by user) and also potential "default" situation in older and unsupported browsers (arguably, there are not much user agents without Web Component support in active use nowadays, but – also arguably – proving there are none is also impossible).

From a quick glance, this project entirely stands on assumption that Web Components work all the time and there is no need (and even way) to compile them to a static boring old-school resilient HTML, for wider compatibility range, or even faster initial load.  Looking at code examples that always use nice clean and convenient declarative components, abstracting the "real" HTML elements away, so without properly working "basically everything", serving `<wa-button href="https://example.com/">Link</wa-button>` to clients will not allow them to navigate anywhere, `<wa-input label="What is your name?"></wa-input>` will not let them input anything, and so forth.

---

As a related side-note: was/is making this project optionally working as a progressive enhancer of "real boring static" HTML elements considered? (I see SSR option was at some point at the table but, as I understand notes around #917 currently is dismissed.)

**Comment by claviska** [comment]
> As a related side-note: was/is making this project optionally working as a progressive enhancer of "real boring static" HTML elements considered?

This was considered early on, but you need to put everything important to PE users in the light DOM. This means a lot of extra markup, a lot of extra wiring (since we can't rely on JS), redundant values that can get out of sync and would need to be watched/mirrored (potentially affecting performance), and a lot more room for user error than what we offer today.

Some examples:

```html
<!-- A simple button becomes less simple -->
<wa-button>
  <button>Click me</button>
</wa-button>

<!-- A link button requires an extra link and redundant attributes to ensure PE* -->
<wa-button href="https://webawesome.com/" target="_blank">
  <a href="href="https://webawesome.com/" target="_blank">Visit Web Awesome</button>
</wa-button>

<!-- Form controls have to have labels and help text manually associated to ensure PE -->
<wa-input name="name">
  <label for="my-input">What is your name?</label>
  <input id="my-input" aria-describedby="my-hint" name="name">
  <div id="my-hint">Please enter your name</div>
</wa-input>
```

<small>

*It could be argued that the component could, when registered, search for the link and mirror the attributes to `<wa-button>` but that would require DOM traversal for every component that behaves this way.

</small>

This also makes it harder to use slots, events, and methods reliably, especially give the extra manual wiring users will inevitably get wrong. Not to mention we wouldn't be able to use ElementInternals to associated form controls since they'd now live in the light DOM — that feels like a huge regression to me.

As someone who champions Web Standards, I used to promote Progressive Enhancement in the 2000s–2010s but the reality is it's a non-issue these days. Very few websites actually function without JavaScript, and folks who turn it off expect a broken experience. It feels like the push for "it must work without JS" is an overcorrection from many years of framework bloat.

The solution isn't no JavaScript, it's less JavaScript.

That said, achieving full PE isn't a goal of this product, especially considering the horrible impact it would have on DX and maintainability (for both us and our users).

We use standard browser APIs and ship only the JS you need for the components you're using. If PE remains a concern, I'd suggest sticking with [Native Styles](https://webawesome.com/docs/utilities/native) which use only native HTML elements and require zero JavaScript.
